lib/auth/keystore/raw.go

```diff
@@ -102,7 +132,61 @@ func (c *rawKeyStore) GetJWTSigner(ca types.CertAuthority) (crypto.Signer, error
 	return nil, trace.NotFound("no JWT key pairs found in CA for %q", ca.GetClusterName())
 }
 
+// NewSSHKeyPair creates and returns a new SSHKeyPair.
+func (c *rawKeyStore) NewSSHKeyPair() (*types.SSHKeyPair, error) {
+	return newSSHKeyPair(c)
+}
+
+// NewTLSKeyPair creates and returns a new TLSKeyPair.
+func (c *rawKeyStore) NewTLSKeyPair(clusterName string) (*types.TLSKeyPair, error) {
+	return newTLSKeyPair(c, clusterName)
+}
+
+// NewJWTKeyPair creates and returns a new JWTKeyPair.
+func (c *rawKeyStore) NewJWTKeyPair() (*types.JWTKeyPair, error) {
+	return newJWTKeyPair(c)
+}
+
 // DeleteKey deletes the given key from the KeyStore. This is a no-op for rawKeyStore.
 func (c *rawKeyStore) DeleteKey(rawKey []byte) error {
 	return nil
 }
+
+func (c *rawKeyStore) keySetHasLocalKeys(keySet types.CAKeySet) bool {
+	for _, sshKeyPair := range keySet.SSH {
+		if sshKeyPair.PrivateKeyType == types.PrivateKeyType_RAW {
+			return true
+		}
+	}
+	for _, tlsKeyPair := range keySet.TLS {
+		if tlsKeyPair.KeyType == types.PrivateKeyType_RAW {
+			return true
+		}
+	}
+	for _, jwtKeyPair := range keySet.JWT {
+		if jwtKeyPair.PrivateKeyType == types.PrivateKeyType_RAW {
+			return true
+		}
+	}
+	return false
+}
+
+// HasLocalActiveKeys returns true if the given CA has any active keys that
+// are usable with this KeyStore.
+func (c *rawKeyStore) HasLocalActiveKeys(ca types.CertAuthority) bool {
+	return c.keySetHasLocalKeys(ca.GetActiveKeys())
+}
+
+// HasLocalAdditionalKeys returns true if the given CA has any additional
+// trusted keys that are usable with this KeyStore.
+func (c *rawKeyStore) HasLocalAdditionalKeys(ca types.CertAuthority) bool {
+	return c.keySetHasLocalKeys(ca.GetAdditionalTrustedKeys())
+}
+
+// DeleteUnusedKeys deletes all keys from the KeyStore if they are:
+// 1. Labeled by this KeyStore when they were created
+// 2. Not included in the argument usedKeys
+// This is a no-op for rawKeyStore.
```

Comment:
to clarify - the raw key will be cleaned up on rotation if all auth servers have HSMs, right?

Reply:
Yes it will be delete from the backend which is the only place it is stored. This no-op is a quirk of `rawKeyStore` not actually holding keys anywhere and just operating directly on what is passed